readme.md
# ft_lab.OmniGraph.GetDateTime

This sample uses OmniGraph to reflect the current time on analog and digital clocks created as 3D models.     
![preview.jpg](./images/preview.jpg)     

This is a sample project in which OmniGraph custom nodes are prepared with a Python Extension to control a pre-prepared 3D model.    

## How to use

I have confirmed that it works with ~~Omniverse Create 2022.3.3~~ USD Composer 2023.2.2(Kit 105.1.2).     
Download and use this repository locally.     

```
[extension]
  [ft_lab.OmniGraph.GetDateTime] ... Extension(OmniGraph Nodes) used in this project

[usds]  sample scene
  [Clock]
    [textures]
    clock.usd
  
  [ClockDigital]
    [textures]
    clock_digital.usd

  clock_stage.usd  ... Open and use this locally.
```

* Assign and activate Extension to Omniverse Create.     
Copy "[ft_lab.OmniGraph.GetDateTime](./extension/ft_lab.OmniGraph.GetDateTime/)" to a folder where Omniverse can find it as an Extension.      
![GetDateTime_extension_01.jpg](./images/GetDateTime_extension_01.jpg)     
* Open "[clock_stage.usd](./usds/clock_stage.usd)" in Omniverse Create.     
References two USD "[clock.usd](./usds/Clock/clock.usd)" and "[clock_digital.usd](./usds/ClockDigital/clock_digital.usd)".

You can now see the current time reflected in the analog and digital clocks.      
![GetDateTime_01.jpg](./images/GetDateTime_01.jpg)     


## Documents

* [Description of OmniGraph nodes](./OmniGraphNodes.md)

## Documents for Development

* [Extension Structure](./docs/ExtensionStructure.md)
  * [GetDateTime](./docs/node_GetDateTime.md)
  * [RotationByTime](./docs/node_RotationByTime.md)
  * [OutputToLCD](./docs/node_OutputToLCD.md)
* [3D Models](./docs/Modeling3D.md)

## Change Log

* [Change Log](./ChangeLog.md)

## License

This software is released under the MIT License, see [LICENSE.txt](./LICENSE.txt).


ChangeLog.md
# Change Log

## December 22, 2023

Fixed in USD Composer 2023.2.2 (Kit.105.1.2)

### xxxxDatabase.py

The icons were not reflected until these two internal versions were updated.     

* GENERATOR_VERSION : (1, 31, 1) -> (1, 41, 3)    
* TARGET_VERSION : (2, 107, 4) -> (2, 139, 12)

## July 11, 2023

Fixed in USD Composer 2023.1.0-beta (Kit.105) from Omniverse Create 2022.3.3 (Kit.104).     

### [RotationByTime.ogn](extension/ft_lab.OmniGraph.GetDateTime/ft_lab/OmniGraph/GetDateTime/nodes/RotationByTime.ogn)

"type": "float3" -> "float[3]"      

### xxxxDatabase.py

* GENERATOR_VERSION : (1, 17, 2) -> (1, 31, 1)    
* TARGET_VERSION : (2, 65, 4) -> (2, 107, 4)



extension/ft_lab.OmniGraph.GetDateTime/ft_lab/OmniGraph/GetDateTime/nodes/OutputToLCD.py
"""
Time output to LCD (hh:mm).
"""
from pxr import Usd, UsdGeom, UsdPhysics, UsdShade, Sdf, Gf, Tf
import numpy as np
import omni.ext

class OutputToLCD:
    @staticmethod
    def compute(db) -> bool:
        try:
            hour   = db.inputs.d1_hour
            minute = db.inputs.d2_minute
            second = db.inputs.d3_second

            # xABCDEFG => 0b01111110 = 0x7e = '0'
            nameList = ["A", "B", "C", "D", "E", "F", "G"]
            numMaskList = [0x7e, 0x30, 0x6d, 0x79, 0x33, 0x5b, 0x5f, 0x70, 0x7f, 0x7b]

            # Get stage.
            stage = omni.usd.get_context().get_stage()

            # Show/hide "AM"
            if db.inputs.c1_amPrim != None and db.inputs.c1_amPrim != "":
                prim = stage.GetPrimAtPath(db.inputs.c1_amPrim)
                if prim.IsValid():
                    primImageable = UsdGeom.Imageable(prim)
                    primImageable.GetVisibilityAttr().Set('inherited' if hour < 12 else 'invisible')

            # Show/hide "PM"
            if db.inputs.c2_pmPrim != None and db.inputs.c2_pmPrim != "":
                prim = stage.GetPrimAtPath(db.inputs.c2_pmPrim)
                if prim.IsValid():
                    primImageable = UsdGeom.Imageable(prim)
                    primImageable.GetVisibilityAttr().Set('inherited' if (hour >= 12) else 'invisible')


            # Hour : 10th digit.
            hour12 = hour if (hour < 12) else (hour - 12)
            if db.inputs.a1_hourNum10Prim != None and db.inputs.a1_hourNum10Prim != "":
                basePrimPath = db.inputs.a1_hourNum10Prim
                shiftV = 0x40
                maskV  = numMaskList[(int)(hour12 / 10) % 10]
                for i in range(7):
                    primPath = f"{basePrimPath}/{nameList[i]}"
                    prim = stage.GetPrimAtPath(primPath)
                    if prim.IsValid():
                        primImageable = UsdGeom.Imageable(prim)
                        primImageable.GetVisibilityAttr().Set('inherited' if ((maskV & shiftV) != 0) else 'invisible')
                    shiftV >>= 1

            # Hour : 1th digit.
            if db.inputs.a2_hourNum1Prim != None and db.inputs.a2_hourNum1Prim != "":
                basePrimPath = db.inputs.a2_hourNum1Prim
                shiftV = 0x40
                maskV  = numMaskList[(int)(hour12) % 10]
                for i in range(7):
                    primPath = f"{basePrimPath}/{nameList[i]}"
                    prim = stage.GetPrimAtPath(primPath)
                    if prim.IsValid():
                        primImageable = UsdGeom.Imageable(prim)
                        primImageable.GetVisibilityAttr().Set('inherited' if ((maskV & shiftV) != 0) else 'invisible')
                    shiftV >>= 1

            # Minute : 10th digit.
            if db.inputs.b1_minuteNum10Prim != None and db.inputs.b1_minuteNum10Prim != "":
                basePrimPath = db.inputs.b1_minuteNum10Prim
                shiftV = 0x40
                maskV  = numMaskList[(int)(minute / 10) % 10]
                for i in range(7):
                    primPath = f"{basePrimPath}/{nameList[i]}"
                    prim = stage.GetPrimAtPath(primPath)
                    if prim.IsValid():
                        primImageable = UsdGeom.Imageable(prim)
                        primImageable.GetVisibilityAttr().Set('inherited' if ((maskV & shiftV) != 0) else 'invisible')
                    shiftV >>= 1

            # Minute : 1th digit.
            if db.inputs.b2_minuteNum1Prim != None and db.inputs.b2_minuteNum1Prim != "":
                basePrimPath = db.inputs.b2_minuteNum1Prim
                shiftV = 0x40
                maskV  = numMaskList[(int)(minute) % 10]
                for i in range(7):
                    primPath = f"{basePrimPath}/{nameList[i]}"
                    prim = stage.GetPrimAtPath(primPath)
                    if prim.IsValid():
                        primImageable = UsdGeom.Imageable(prim)
                        primImageable.GetVisibilityAttr().Set('inherited' if ((maskV & shiftV) != 0) else 'invisible')
                    shiftV >>= 1

        except TypeError as error:
            db.log_error(f"Processing failed : {error}")
            return False
    
        return True




extension/ft_lab.OmniGraph.GetDateTime/ft_lab/OmniGraph/GetDateTime/nodes/RotationByTime.py
"""
Rotation by time.
"""
import numpy as np
import omni.ext

class RotationByTime:
    @staticmethod
    def compute(db) -> bool:
        try:
            # Calculate clock rotation from seconds.
            if db.inputs.a2_rotationAxis >= 0 and db.inputs.a2_rotationAxis <= 2:
                v = db.outputs.a3_secondRotateXYZ
                v[0] = db.inputs.a1_defaultRotateXYZ[0]
                v[1] = db.inputs.a1_defaultRotateXYZ[1]
                v[2] = db.inputs.a1_defaultRotateXYZ[2]
                v[db.inputs.a2_rotationAxis] = ((float)(db.inputs.b3_second) / 60.0) * 360.0

            # Calculate clock rotation from minutes.
            if db.inputs.a2_rotationAxis >= 0 and db.inputs.a2_rotationAxis <= 2:
                v = db.outputs.a2_minuteRotateXYZ
                v[0] = db.inputs.a1_defaultRotateXYZ[0]
                v[1] = db.inputs.a1_defaultRotateXYZ[1]
                v[2] = db.inputs.a1_defaultRotateXYZ[2]
                v[db.inputs.a2_rotationAxis] = ((float)(db.inputs.b2_minute * 60.0 + db.inputs.b3_second) / (60.0 * 60.0)) * 360.0

            # Calculate clock rotation from hours.
            if db.inputs.a2_rotationAxis >= 0 and db.inputs.a2_rotationAxis <= 2:
                v = db.outputs.a1_hourRotateXYZ
                v[0] = db.inputs.a1_defaultRotateXYZ[0]
                v[1] = db.inputs.a1_defaultRotateXYZ[1]
                v[2] = db.inputs.a1_defaultRotateXYZ[2]
                v[db.inputs.a2_rotationAxis] = ((float)(db.inputs.b1_hour * 60.0 + db.inputs.b2_minute) / (60.0 * 24.0)) * 360.0 * 2.0

        except TypeError as error:
            db.log_error(f"Processing failed : {error}")
            return False
    
        return True




extension/ft_lab.OmniGraph.GetDateTime/ft_lab/OmniGraph/GetDateTime/ogn/GetDateTimeDatabase.py
import omni.graph.core as og
import omni.graph.core._omni_graph_core as _og
import omni.graph.tools.ogn as ogn
import numpy
import sys
import traceback
import carb

class GetDateTimeDatabase(og.Database):
    """Helper class providing simplified access to data on nodes of type ft_lab.OmniGraph.GetDateTime.GetDateTime

    Class Members:
        node: Node being evaluated

    Attribute Value Properties:
        Inputs:
        Outputs:
            outputs.a1_year
            outputs.a2_month
            outputs.a3_day
            outputs.b1_hour
            outputs.b2_minute
            outputs.b3_second
    """

    # Omniverse Create 2022.3.3 (Kit.104)
    #GENERATOR_VERSION = (1, 17, 2)
    #TARGET_VERSION = (2, 65, 4)

    # Imprint the generator and target ABI versions in the file for JIT generation
    # USD Composer 2023.2.2 (Kit.105.1.2)
    GENERATOR_VERSION = (1, 41, 3)
    TARGET_VERSION = (2, 139, 12)

    # This is an internal object that provides per-class storage of a per-node data dictionary
    PER_NODE_DATA = {}

    INTERFACE = og.Database._get_interface([
        ('outputs:a1_year', 'int', 0, 'Year', 'output year', {ogn.MetadataKeys.DEFAULT: '2000'}, True, 0, False, ''),
        ('outputs:a2_month', 'int', 0, 'Month', 'output month', {ogn.MetadataKeys.DEFAULT: '1'}, True, 0, False, ''),
        ('outputs:a3_day', 'int', 0, 'Day', 'output day', {ogn.MetadataKeys.DEFAULT: '1'}, True, 0, False, ''),
        ('outputs:b1_hour', 'int', 0, 'Hour', 'output hour', {ogn.MetadataKeys.DEFAULT: '1'}, True, 0, False, ''),
        ('outputs:b2_minute', 'int', 0, 'Minute', 'output minute', {ogn.MetadataKeys.DEFAULT: '0'}, True, 0, False, ''),
        ('outputs:b3_second', 'int', 0, 'Second', 'output second', {ogn.MetadataKeys.DEFAULT: '0'}, True, 0, False, ''),
    ])

    # ----------------------------------------------------.
    # Processing Output Parameter.
    # ----------------------------------------------------.
    class ValuesForOutputs(og.DynamicAttributeAccess):
        LOCAL_PROPERTY_NAMES = { "a1_year", "a2_month", "a3_day", "b1_hour", "b2_month", "b3_second" }
        """Helper class that creates natural hierarchical access to output attributes"""
        def __init__(self, node: og.Node, attributes, dynamic_attributes: og.DynamicAttributeInterface):
            """Initialize simplified access for the attribute data"""
            context = node.get_graph().get_default_graph_context()
            super().__init__(context, node, attributes, dynamic_attributes)
            self._batchedWriteValues = { }

        @property
        def a1_year(self):
            value = self._batchedWriteValues.get(self._attributes.a1_year)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a1_year)
                return data_view.get()

        @a1_year.setter
        def a1_year(self, value):
            self._batchedWriteValues[self._attributes.a1_year] = value

        @property
        def a2_month(self):
            value = self._batchedWriteValues.get(self._attributes.a2_month)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a2_month)
                return data_view.get()

        @a2_month.setter
        def a2_month(self, value):
            self._batchedWriteValues[self._attributes.a2_month] = value

        @property
        def a3_day(self):
            value = self._batchedWriteValues.get(self._attributes.a3_day)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a3_day)
                return data_view.get()

        @a3_day.setter
        def a3_day(self, value):
            self._batchedWriteValues[self._attributes.a3_day] = value

        @property
        def b1_hour(self):
            value = self._batchedWriteValues.get(self._attributes.b1_hour)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.b1_hour)
                return data_view.get()

        @b1_hour.setter
        def b1_hour(self, value):
            self._batchedWriteValues[self._attributes.b1_hour] = value

        @property
        def b2_minute(self):
            value = self._batchedWriteValues.get(self._attributes.b2_minute)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.b2_minute)
                return data_view.get()

        @b2_minute.setter
        def b2_minute(self, value):
            self._batchedWriteValues[self._attributes.b2_minute] = value

        @property
        def b3_second(self):
            value = self._batchedWriteValues.get(self._attributes.b3_second)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.b3_second)
                return data_view.get()

        @b3_second.setter
        def b3_second(self, value):
            self._batchedWriteValues[self._attributes.b3_second] = value


        def __getattr__(self, item: str):
            if item in self.LOCAL_PROPERTY_NAMES:
                return object.__getattribute__(self, item)
            else:
                return super().__getattr__(item)

        def __setattr__(self, item: str, new_value):
            if item in self.LOCAL_PROPERTY_NAMES:
                object.__setattr__(self, item, new_value)
            else:
                super().__setattr__(item, new_value)

        def _commit(self):
            _og._commit_output_attributes_data(self._batchedWriteValues)
            self._batchedWriteValues = { }

    class ValuesForState(og.DynamicAttributeAccess):
        """Helper class that creates natural hierarchical access to state attributes"""
        def __init__(self, node: og.Node, attributes, dynamic_attributes: og.DynamicAttributeInterface):
            """Initialize simplified access for the attribute data"""
            context = node.get_graph().get_default_graph_context()
            super().__init__(context, node, attributes, dynamic_attributes)

    def __init__(self, node):
        super().__init__(node)

        dynamic_attributes = self.dynamic_attribute_data(node, og.AttributePortType.ATTRIBUTE_PORT_TYPE_OUTPUT)
        self.outputs = GetDateTimeDatabase.ValuesForOutputs(node, self.attributes.outputs, dynamic_attributes)

        dynamic_attributes = self.dynamic_attribute_data(node, og.AttributePortType.ATTRIBUTE_PORT_TYPE_STATE)
        self.state = GetDateTimeDatabase.ValuesForState(node, self.attributes.state, dynamic_attributes)

    # ----------------------------------------------------.
    # Class defining the ABI interface for the node type.
    # ----------------------------------------------------.
    class abi:
        @staticmethod
        def get_node_type():
            get_node_type_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'get_node_type', None)
            if callable(get_node_type_function):
                return get_node_type_function()
            return 'ft_lab.OmniGraph.GetDateTime.GetDateTime'
        
        @staticmethod
        def compute(context, node):
            def database_valid():
                return True
            try:
                per_node_data = GetDateTimeDatabase.PER_NODE_DATA[node.node_id()]
                db = per_node_data.get('_db')
                if db is None:
                    db = GetDateTimeDatabase(node)
                    per_node_data['_db'] = db
                if not database_valid():
                    per_node_data['_db'] = None
                    return False
            except:
                db = GetDateTimeDatabase(node)

            try:
                compute_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'compute', None)
                if callable(compute_function) and compute_function.__code__.co_argcount > 1:
                    return compute_function(context, node)

                with og.in_compute():
                    return GetDateTimeDatabase.NODE_TYPE_CLASS.compute(db)
            except Exception as error:
                stack_trace = "".join(traceback.format_tb(sys.exc_info()[2].tb_next))
                db.log_error(f'Assertion raised in compute - {error}\n{stack_trace}', add_context=False)
            finally:
                db.outputs._commit()
            return False
        @staticmethod
        def initialize(context, node):
            GetDateTimeDatabase._initialize_per_node_data(node)
            initialize_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'initialize', None)
            if callable(initialize_function):
                initialize_function(context, node)
        @staticmethod
        def release(node):
            release_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'release', None)
            if callable(release_function):
                release_function(node)
            GetDateTimeDatabase._release_per_node_data(node)
        @staticmethod
        def update_node_version(context, node, old_version, new_version):
            update_node_version_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'update_node_version', None)
            if callable(update_node_version_function):
                return update_node_version_function(context, node, old_version, new_version)
            return False
        @staticmethod
        def initialize_type(node_type):
            initialize_type_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'initialize_type', None)
            needs_initializing = True
            if callable(initialize_type_function):
                needs_initializing = initialize_type_function(node_type)
            if needs_initializing:
                node_type.set_metadata(ogn.MetadataKeys.EXTENSION, "ft_lab.OmniGraph.GetDateTime")
                node_type.set_metadata(ogn.MetadataKeys.UI_NAME, "Get DateTime")
                node_type.set_metadata(ogn.MetadataKeys.CATEGORIES, "examples")
                node_type.set_metadata(ogn.MetadataKeys.DESCRIPTION, "Get current date and time")
                node_type.set_metadata(ogn.MetadataKeys.LANGUAGE, "Python")

                # Set Icon(svg).
                icon_path = carb.tokens.get_tokens_interface().resolve("${ft_lab.OmniGraph.GetDateTime}")
                icon_path = icon_path + '/' + "data/icons/ft_lab.OmniGraph.GetDateTime.icon.svg"
                node_type.set_metadata(ogn.MetadataKeys.ICON_PATH, icon_path)

                GetDateTimeDatabase.INTERFACE.add_to_node_type(node_type)
        @staticmethod
        def on_connection_type_resolve(node):
            on_connection_type_resolve_function = getattr(GetDateTimeDatabase.NODE_TYPE_CLASS, 'on_connection_type_resolve', None)
            if callable(on_connection_type_resolve_function):
                on_connection_type_resolve_function(node)
    NODE_TYPE_CLASS = None

    @staticmethod
    def register(node_type_class):
        GetDateTimeDatabase.NODE_TYPE_CLASS = node_type_class
        og.register_node_type(GetDateTimeDatabase.abi, 1)

    @staticmethod
    def deregister():
        og.deregister_node_type("ft_lab.OmniGraph.GetDateTime.GetDateTime")


extension/ft_lab.OmniGraph.GetDateTime/ft_lab/OmniGraph/GetDateTime/ogn/RotationByTimeDatabase.py
import omni.graph.core as og
import omni.graph.core._omni_graph_core as _og
import omni.graph.tools.ogn as ogn
import numpy
import sys
import traceback
import carb

class RotationByTimeDatabase(og.Database):
    """Helper class providing simplified access to data on nodes of type ft_lab.OmniGraph.GetDateTime.RotationByTime

    Class Members:
        node: Node being evaluated

    Attribute Value Properties:
        Inputs:
            inputs.a1_defaultRotateXYZ
            inputs.a2_rotationAxis
            inputs.b1_hour
            inputs.b2_minute
            inputs.b3_second
        Outputs:
            outputs.a1_hourRotateXYZ
            outputs.a2_minuteRotateXYZ
            outputs.a3_secondRotateXYZ
    """

    # Omniverse Create 2022.3.3 (Kit.104)
    #GENERATOR_VERSION = (1, 17, 2)
    #TARGET_VERSION = (2, 65, 4)

    # Imprint the generator and target ABI versions in the file for JIT generation
    # USD Composer 2023.2.2 (Kit.105.1.2)
    GENERATOR_VERSION = (1, 41, 3)
    TARGET_VERSION = (2, 139, 12)

    # This is an internal object that provides per-class storage of a per-node data dictionary
    PER_NODE_DATA = {}

    INTERFACE = og.Database._get_interface([
        ('inputs:a1_defaultRotateXYZ', 'float[3]', 0, 'Default RotateXYZ', 'Default rotateXYZ', {}, True, None, False, ''),
        ('inputs:a2_rotationAxis', 'int', 0, 'Rotation Axis', 'Rotation axis (0:X, 1:Y, 2:Z)', {}, True, None, False, ''),
        ('inputs:b1_hour', 'int', 0, 'Hour', 'Hour', {ogn.MetadataKeys.DEFAULT: '0'}, True, 0, False, ''),
        ('inputs:b2_minute', 'int', 0, 'Minute', 'Minute', {ogn.MetadataKeys.DEFAULT: '0'}, True, 0, False, ''),
        ('inputs:b3_second', 'int', 0, 'Second', 'Second', {ogn.MetadataKeys.DEFAULT: '0'}, True, 0, False, ''),
        ('outputs:a1_hourRotateXYZ', 'float[3]', 0, 'Hour RotateXYZ', 'Hour RotateXYZ', {}, True, None, False, ''),
        ('outputs:a2_minuteRotateXYZ', 'float[3]', 0, 'Minute RotateXYZ', 'Minute RotateXYZ', {}, True, None, False, ''),
        ('outputs:a3_secondRotateXYZ', 'float[3]', 0, 'Second RotateXYZ', 'Second RotateXYZ', {}, True, None, False, ''),
    ])

    # ----------------------------------------------------.
    # Processing Input Parameters.
    # ----------------------------------------------------.
    class ValuesForInputs(og.DynamicAttributeAccess):
        LOCAL_PROPERTY_NAMES = {"a1_defaultRotateXYZ", "a2_rotationAxis", "b1_hour", "b2_minute", "b3_second"}
        """Helper class that creates natural hierarchical access to input attributes"""
        def __init__(self, node: og.Node, attributes, dynamic_attributes: og.DynamicAttributeInterface):
            """Initialize simplified access for the attribute data"""
            context = node.get_graph().get_default_graph_context()
            super().__init__(context, node, attributes, dynamic_attributes)
            self._batchedReadAttributes = [self._attributes.a1_defaultRotateXYZ, self._attributes.a2_rotationAxis, self._attributes.b1_hour, self._attributes.b2_minute, self._attributes.b3_second]
            self._batchedReadValues = [[0.0, 0.0, 0.0], 0, 0, 0, 0]

        @property
        def a1_defaultRotateXYZ(self):
            return self._batchedReadValues[0]

        @a1_defaultRotateXYZ.setter
        def a1_defaultRotateXYZ(self, value):
            self._batchedReadValues[0] = value

        @property
        def a2_rotationAxis(self):
            return self._batchedReadValues[1]

        @a2_rotationAxis.setter
        def a2_rotationAxis(self, value):
            self._batchedReadValues[1] = value

        @property
        def b1_hour(self):
            return self._batchedReadValues[2]

        @b1_hour.setter
        def b1_hour(self, value):
            self._batchedReadValues[2] = value

        @property
        def b2_minute(self):
            return self._batchedReadValues[3]

        @b2_minute.setter
        def b2_minute(self, value):
            self._batchedReadValues[3] = value

        @property
        def b3_second(self):
            return self._batchedReadValues[4]

        @b3_second.setter
        def b3_second(self, value):
            self._batchedReadValues[4] = value

        def __getattr__(self, item: str):
            if item in self.LOCAL_PROPERTY_NAMES:
                return object.__getattribute__(self, item)
            else:
                return super().__getattr__(item)

        def __setattr__(self, item: str, new_value):
            if item in self.LOCAL_PROPERTY_NAMES:
                object.__setattr__(self, item, new_value)
            else:
                super().__setattr__(item, new_value)

        def _prefetch(self):
            readAttributes = self._batchedReadAttributes
            newValues = _og._prefetch_input_attributes_data(readAttributes)
            if len(readAttributes) == len(newValues):
                self._batchedReadValues = newValues

    # ----------------------------------------------------.
    # Processing Output Parameter.
    # ----------------------------------------------------.
    class ValuesForOutputs(og.DynamicAttributeAccess):
        LOCAL_PROPERTY_NAMES = { "a1_hourRotateXYZ", "a2_minuiteRotateXYZ", "a3_secondRotateXYZ" }
        """Helper class that creates natural hierarchical access to output attributes"""
        def __init__(self, node: og.Node, attributes, dynamic_attributes: og.DynamicAttributeInterface):
            """Initialize simplified access for the attribute data"""
            context = node.get_graph().get_default_graph_context()
            super().__init__(context, node, attributes, dynamic_attributes)
            self._batchedWriteValues = { }

        @property
        def a1_hourRotateXYZ(self):
            value = self._batchedWriteValues.get(self._attributes.a1_hourRotateXYZ)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a1_hourRotateXYZ)
                return data_view.get()

        @a1_hourRotateXYZ.setter
        def a1_hourRotateXYZ(self, value):
            self._batchedWriteValues[self._attributes.a1_hourRotateXYZ] = value

        @property
        def a2_minuteRotateXYZ(self):
            value = self._batchedWriteValues.get(self._attributes.a2_minuteRotateXYZ)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a2_minuteRotateXYZ)
                return data_view.get()

        @a2_minuteRotateXYZ.setter
        def a2_minuteRotateXYZ(self, value):
            self._batchedWriteValues[self._attributes.a2_minuteRotateXYZ] = value

        @property
        def a3_secondRotateXYZ(self):
            value = self._batchedWriteValues.get(self._attributes.a3_secondRotateXYZ)
            if value:
                return value
            else:
                data_view = og.AttributeValueHelper(self._attributes.a3_secondRotateXYZ)
                return data_view.get()

        @a3_secondRotateXYZ.setter
        def a3_secondRotateXYZ(self, value):
            self._batchedWriteValues[self._attributes.a3_secondRotateXYZ] = value

        def __getattr__(self, item: str):
            if item in self.LOCAL_PROPERTY_NAMES:
                return object.__getattribute__(self, item)
            else:
                return super().__getattr__(item)

        def __setattr__(self, item: str, new_value):
            if item in self.LOCAL_PROPERTY_NAMES:
                object.__setattr__(self, item, new_value)
            else:
                super().__setattr__(item, new_value)

        def _commit(self):
            _og._commit_output_attributes_data(self._batchedWriteValues)
            self._batchedWriteValues = { }

    class ValuesForState(og.DynamicAttributeAccess):
        """Helper class that creates natural hierarchical access to state attributes"""
        def __init__(self, node: og.Node, attributes, dynamic_attributes: og.DynamicAttributeInterface):
            """Initialize simplified access for the attribute data"""
            context = node.get_graph().get_default_graph_context()
            super().__init__(context, node, attributes, dynamic_attributes)

    def __init__(self, node):
        super().__init__(node)

        dynamic_attributes = self.dynamic_attribute_data(node, og.AttributePortType.ATTRIBUTE_PORT_TYPE_INPUT)
        self.inputs = RotationByTimeDatabase.ValuesForInputs(node, self.attributes.inputs, dynamic_attributes)

        dynamic_attributes = self.dynamic_attribute_data(node, og.AttributePortType.ATTRIBUTE_PORT_TYPE_OUTPUT)
        self.outputs = RotationByTimeDatabase.ValuesForOutputs(node, self.attributes.outputs, dynamic_attributes)

        dynamic_attributes = self.dynamic_attribute_data(node, og.AttributePortType.ATTRIBUTE_PORT_TYPE_STATE)
        self.state = RotationByTimeDatabase.ValuesForState(node, self.attributes.state, dynamic_attributes)

    # ----------------------------------------------------.
    # Class defining the ABI interface for the node type.
    # ----------------------------------------------------.
    class abi:
        @staticmethod
        def get_node_type():
            get_node_type_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'get_node_type', None)
            if callable(get_node_type_function):
                return get_node_type_function()
            return 'ft_lab.OmniGraph.GetDateTime.RotationByTime'
        
        @staticmethod
        def compute(context, node):
            def database_valid():
                return True
            try:
                per_node_data = RotationByTimeDatabase.PER_NODE_DATA[node.node_id()]
                db = per_node_data.get('_db')
                if db is None:
                    db = RotationByTimeDatabase(node)
                    per_node_data['_db'] = db
                if not database_valid():
                    per_node_data['_db'] = None
                    return False
            except:
                db = RotationByTimeDatabase(node)

            try:
                compute_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'compute', None)
                if callable(compute_function) and compute_function.__code__.co_argcount > 1:
                    return compute_function(context, node)

                db.inputs._prefetch()
                db.inputs._setting_locked = True
                with og.in_compute():
                    return RotationByTimeDatabase.NODE_TYPE_CLASS.compute(db)
            except Exception as error:
                stack_trace = "".join(traceback.format_tb(sys.exc_info()[2].tb_next))
                db.log_error(f'Assertion raised in compute - {error}\n{stack_trace}', add_context=False)
            finally:
                db.inputs._setting_locked = False
                db.outputs._commit()
            return False
        @staticmethod
        def initialize(context, node):
            RotationByTimeDatabase._initialize_per_node_data(node)
            initialize_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'initialize', None)
            if callable(initialize_function):
                initialize_function(context, node)
        @staticmethod
        def release(node):
            release_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'release', None)
            if callable(release_function):
                release_function(node)
            RotationByTimeDatabase._release_per_node_data(node)
        @staticmethod
        def update_node_version(context, node, old_version, new_version):
            update_node_version_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'update_node_version', None)
            if callable(update_node_version_function):
                return update_node_version_function(context, node, old_version, new_version)
            return False
        @staticmethod
        def initialize_type(node_type):
            initialize_type_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'initialize_type', None)
            needs_initializing = True
            if callable(initialize_type_function):
                needs_initializing = initialize_type_function(node_type)
            if needs_initializing:
                node_type.set_metadata(ogn.MetadataKeys.EXTENSION, "ft_lab.OmniGraph.GetDateTime")
                node_type.set_metadata(ogn.MetadataKeys.UI_NAME, "Rotation By Time")
                node_type.set_metadata(ogn.MetadataKeys.CATEGORIES, "examples")
                node_type.set_metadata(ogn.MetadataKeys.DESCRIPTION, "Rotation By Time")
                node_type.set_metadata(ogn.MetadataKeys.LANGUAGE, "Python")

                # Set Icon(svg).
                icon_path = carb.tokens.get_tokens_interface().resolve("${ft_lab.OmniGraph.GetDateTime}")
                icon_path = icon_path + '/' + "data/icons/ft_lab.OmniGraph.GetDateTime.rotationByTimeIcon.svg"
                node_type.set_metadata(ogn.MetadataKeys.ICON_PATH, icon_path)

                RotationByTimeDatabase.INTERFACE.add_to_node_type(node_type)
        @staticmethod
        def on_connection_type_resolve(node):
            on_connection_type_resolve_function = getattr(RotationByTimeDatabase.NODE_TYPE_CLASS, 'on_connection_type_resolve', None)
            if callable(on_connection_type_resolve_function):
                on_connection_type_resolve_function(node)
    NODE_TYPE_CLASS = None

    @staticmethod
    def register(node_type_class):
        RotationByTimeDatabase.NODE_TYPE_CLASS = node_type_class
        og.register_node_type(RotationByTimeDatabase.abi, 1)

    @staticmethod
    def deregister():
        og.deregister_node_type("ft_lab.OmniGraph.GetDateTime.RotationByTime")
